core/domain/user_services.py

```diff
@@ -499,6 +499,139 @@ def get_user_settings_by_auth_id(
+            which validation needs to be done.
+
+    Raises:
+        Exception. The user inside user group does not exist.
+    """
+    all_users_usernames = get_all_users_usernames()
+    for user in user_group_users:
+        if not user in all_users_usernames:
+            raise Exception(
+                f'The user {user} of user-group {user_group_name} '
+                'does not exist.')
+
+
+def create_new_user_group(
+    user_group_name: str, users: List[str]
+) -> user_domain.UserGroup:
+    """Create new user group.
+
+    Args:
+        user_group_name: str. The name of the user group.
+        users: List[str]. The users associated with the user group.
+
+    Returns:
+        UserGroup. The new user group.
+    """
+    if len(users) > 0:
+        _check_if_username_is_valid(user_group_name, users)
+    user_group_id = user_models.UserGroupModel.get_new_id('')
+    user_group_dict: user_domain.UserGroupDict = {
+        'user_group_id': user_group_id,
+        'name': user_group_name,
+        'users': users
+    }
+    user_group = user_domain.UserGroup.from_dict(user_group_dict)
+    user_group.validate()
+
+    user_models.UserGroupModel(
+        id=user_group.user_group_id,
+        name=user_group.name,
+        users=user_group.users
+    ).put()
+    return user_group
+
+
+def delete_user_group(user_group_id: str) -> None:
+    """Delete the user group with specified id.
+
+    Args:
+        user_group_id: str. The id of the user group to delete.
+
+    Raises:
+        Exception. The user group trying to delete does not exist.
+    """
+    user_group_model = user_models.UserGroupModel.get(
+        user_group_id, strict=False)
+    if user_group_model is None:
+        raise Exception(f'User group with id {user_group_id} does not exist.')
+    assert user_group_model is not None
+    user_group_model.delete()
+
+
+def update_user_group(
+    user_group_id: str,
+    user_group_name: str,
+    user_group_users: List[str]
+) -> None:
+    """Updates the user group.
+
+    Args:
+        user_group_id: str. The user group id.
+        user_group_name: str. The new name of the user group if needs to
+            be updated else old name of the user group.
+        user_group_users: List[str]. The list of users for the specified
+            user group.
+
+    Raises:
+        Exception. The user group trying to update does not exist.
+    """
+    user_group_model = user_models.UserGroupModel.get(
+        user_group_id, strict=False)
+    if user_group_model is None:
+        raise Exception(f'User group {user_group_name} does not exist.')
+    assert user_group_model is not None
+
+    if len(user_group_users) > 0:
+        _check_if_username_is_valid(user_group_name, user_group_users)
+    if len(user_group_users) != len(set(user_group_users)):
+        raise Exception(
+            f'Users list of user-group {user_group_name} contains '
+            'duplicates.'
```

Comment:
State, in the error message, what the duplicates are. (You can perhaps do that by subtracting the set from the original list and then applying set() to the result to make the elements unique, though there might be simpler ways.)

Reply:
Got it, updated.